beacon-chain/db/slasherkv/pruning.go

```diff
@@ -11,199 +11,153 @@ import (
-				if !uint64PrefixLessThan(k, encodedEndPruneSlot) {
-					epochAtCursor = endEpoch
-					return nil
-				}
-				epochAtCursor = helpers.SlotToEpoch(slotFromProposalKey(k))
-
-				// Proposals in the database look like this:
-				//  (slot ++ validatorIndex) => encode(proposal)
-				// so it is possible we have a few adjacent objects that have the same slot, such as
-				//  (slot = 3 ++ validatorIndex = 0) => ...
-				//  (slot = 3 ++ validatorIndex = 1) => ...
-				//  (slot = 3 ++ validatorIndex = 2) => ...
-				// so we only mark an epoch as pruned if the epoch of the current object
-				// under the cursor has changed.
-				if err := proposalBkt.Delete(k); err != nil {
-					return err
-				}
-				if epochAtCursor == 0 {
-					epochsPruned = 1
-				} else if epochAtCursor > lastPrunedEpoch {
-					epochsPruned++
-					lastPrunedEpoch = epochAtCursor
-				}
-				slasherProposalsPrunedTotal.Inc()
-
-				// If we have pruned N epochs in this pruning iteration,
-				// we exit from the bolt transaction.
-				if epochsPruned >= pruningEpochIncrements {
-					return nil
-				}
+	if err = s.db.Update(func(tx *bolt.Tx) error {
+		signingRootsBkt := tx.Bucket(attestationDataRootsBucket)
+		attRecordsBkt := tx.Bucket(attestationRecordsBucket)
+		c := signingRootsBkt.Cursor()
+
+		// We begin a pruning iteration starting from the first item in the bucket.
+		for k, v := c.First(); k != nil; k, v = c.Next() {
+			// We check the epoch from the current key in the database.
+			// If we have hit an epoch that is greater than the end epoch of the pruning process,
+			// we then completely exit the process as we are done.
+			if uint64PrefixGreaterThan(k, encodedEndPruneEpoch) {
+				return nil
 			}
-			return nil
-		}); err != nil {
-			return err
+
+			// Attestation in the database look like this:
+			//  (target_epoch ++ _) => encode(attestation)
+			// so it is possible we have a few adjacent objects that have the same slot, such as
+			//  (target_epoch = 3 ++ _) => encode(attestation)
+			if err := signingRootsBkt.Delete(k); err != nil {
+				return err
+			}
+			if err := attRecordsBkt.Delete(v); err != nil {
+				return err
+			}
+			slasherAttestationsPrunedTotal.Inc()
+			numPruned++
 		}
+		return nil
+	}); err != nil {
+		return
 	}
-	return nil
+	return
 }
 
-// PruneAttestations prunes all proposal data older than historyLength.
-func (s *Store) PruneAttestations(
-	ctx context.Context, currentEpoch, pruningEpochIncrements, historyLength types.Epoch,
-) error {
-	if currentEpoch < historyLength {
-		log.Debugf("Current epoch %d < history length %d, nothing to prune", currentEpoch, historyLength)
-		return nil
+// PruneProposalsAtEpoch deletes all proposals from the slasher DB with epoch
+// less than or equal to the specified epoch.
+func (s *Store) PruneProposalsAtEpoch(
+	ctx context.Context, maxEpoch types.Epoch,
+) (numPruned uint, err error) {
+	var endPruneSlot types.Slot
+	endPruneSlot, err = helpers.StartSlot(maxEpoch)
```

Comment:
I think this needs to be `helpers.StartSlot(maxEpoch) - 1`

Reply:
@terencechain maybe the func name can be improved, but it relies on pruning up to and including the max epoch specified to it

```
// PruneProposalsAtEpoch deletes all proposals from the slasher DB with epoch
// less than or equal to the specified epoch.
```